community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/DirectoryFactory.java

```diff
@@ -52,84 +51,140 @@
         }
+    }
+
+    /**
+     * Serves reads directly from file system, via heap buffers.
+     */
+    final class HeapDirectoryFactory implements DirectoryFactory
+    {
+        @SuppressWarnings( "ResultOfMethodCallIgnored" )
+        @Override
+        public Directory open( File dir ) throws IOException
+        {
+            Files.createDirectories( dir.toPath() );
+            return new NIOFSDirectory( dir.toPath() );
+        }
 
         @Override
-        public void dumpToZip( ZipOutputStream zip, byte[] scratchPad )
+        public void close()
         {
-            // do nothing
+            // No resources to release. This method only exists as a hook for test implementations.
         }
-    };
+    }
 
-    final class InMemoryDirectoryFactory implements DirectoryFactory
+    /**
+     * Produces directories that read via the provided page cache, meaning, for the read path, this
+     * has good performance and fixed memory use. The write path could still use some optimization.
+     */
+    final class PagedDirectoryFactory implements DirectoryFactory
     {
-        private final Map<File, RAMDirectory> directories = new HashMap<File, RAMDirectory>( );
+        private final PageCache pageCache;
+
+        public PagedDirectoryFactory( PageCache pageCache )
+        {
+            this.pageCache = pageCache;
+        }
 
         @Override
-        public synchronized Directory open( File dir ) throws IOException
+        public Directory open( File dir ) throws IOException
         {
-            if ( !directories.containsKey( dir ) )
-            {
-                directories.put( dir, new RAMDirectory() );
-            }
-            return new UncloseableDirectory( directories.get( dir ) );
+            return new PagedDirectory( dir.toPath(), pageCache );
         }
 
         @Override
-        public synchronized void close()
+        public void close()
         {
-            for ( RAMDirectory ramDirectory : directories.values() )
-            {
-                ramDirectory.close();
-            }
-            directories.clear();
+
+        }
+    }
+
+    /**
+     * Lucene creates one immutable file for every commit - lots of tiny files. Over time, it merges these files into
+     * larger ones, but the point remains that lots of tiny files are part of the basic operation of the index.
+     * <p>
+     * This class limits the actual physical file creation by caching the tiny files in heap buffers, attempting
+     * to merge them together into larger files, that in turn actually get written to disk.
+     */
+    final class NRTCachingDirectoryFactory implements DirectoryFactory
+    {
+        // Its unclear to me that this actually benefits us - we refresh readers constantly, wouldn't
+        // that force this to flush all the time? This adds substantial heap load by allocating
+        // buffers for every index write on the heap - does it bring a clear benefit?
```

Comment:
yes it is, using NRT for us gives a lot of perf benefits

Reply:
got it, removing this comment